precompiles/collective/src/lib.rs

```diff
@@ -0,0 +1,281 @@
+			.map_err(|_| RevertReason::custom("Failed to decode proposal").in_field("proposal"))?
+			.into();
+		let proposal = Box::new(proposal);
+
+		let origin = Runtime::AddressMapping::into_account_id(handle.context().caller);
+		RuntimeHelper::<Runtime>::try_dispatch(
+			handle,
+			Some(origin).into(),
+			pallet_collective::Call::<Runtime, Instance>::execute {
+				proposal,
+				length_bound: proposal_length,
+			},
+		)?;
+
+		let log = log2(
+			handle.context().address,
+			SELECTOR_LOG_EXECUTED,
+			proposal_hash,
+			Vec::new(),
+		);
+
+		handle.record_log_costs(&[&log])?;
+		log.record(handle)?;
+
+		Ok(succeed(&[]))
+	}
+
+	fn propose(handle: &mut impl PrecompileHandle) -> EvmResult<PrecompileOutput> {
+		handle.record_cost(RuntimeHelper::<Runtime>::db_read_gas_cost())?;
+
+		read_args!(handle, {
+			threshold: u32,
+			proposal: BoundedBytes<GetProposalLimit>
+		});
+
+		let proposal: Vec<_> = proposal.into_vec();
+		let proposal_length: u32 = proposal.len().try_into().map_err(|_| {
+			RevertReason::value_is_too_large("uint32")
+				.in_field("length")
+				.in_field("proposal")
+		})?;
+
+		let proposal_index = pallet_collective::Pallet::<Runtime, Instance>::proposal_count();
+		let proposal_hash: H256 = Runtime::Hashing::hash(&proposal).into();
+		let proposal = Runtime::Call::decode(&mut &*proposal)
+			.map_err(|_| RevertReason::custom("Failed to decode proposal").in_field("proposal"))?
+			.into();
+		let proposal = Box::new(proposal);
+
+		{
+			let origin = Runtime::AddressMapping::into_account_id(handle.context().caller);
+			RuntimeHelper::<Runtime>::try_dispatch(
+				handle,
+				Some(origin).into(),
+				pallet_collective::Call::<Runtime, Instance>::propose {
+					threshold,
+					proposal,
+					length_bound: proposal_length,
+				},
+			)?;
+		}
+
+		// In pallet_collective a threshold < 2 means the proposal has been
+		// executed directly.
+		let log = if threshold < 2 {
+			log2(
+				handle.context().address,
+				SELECTOR_LOG_EXECUTED,
+				proposal_hash,
+				Vec::new(),
+			)
+		} else {
+			log4(
+				handle.context().address,
+				SELECTOR_LOG_PROPOSED,
+				handle.context().caller,
+				H256::from_slice(&EvmDataWriter::new().write(proposal_index).build()),
+				proposal_hash,
+				EvmDataWriter::new().write(threshold).build(),
+			)
+		};
+
+		handle.record_log_costs(&[&log])?;
+		log.record(handle)?;
+
+		Ok(succeed(&[]))
+	}
+
+	fn vote(handle: &mut impl PrecompileHandle) -> EvmResult<PrecompileOutput> {
+		read_args!(handle, {
+			proposal_hash: H256,
+			proposal_index: u32,
+			approve: bool
+		});
+
+		let origin = Runtime::AddressMapping::into_account_id(handle.context().caller);
+		RuntimeHelper::<Runtime>::try_dispatch(
+			handle,
+			Some(origin).into(),
+			pallet_collective::Call::<Runtime, Instance>::vote {
+				proposal: proposal_hash.into(),
+				index: proposal_index,
+				approve,
+			},
+		)?;
+
+		// TODO: Since we cannot access ayes/nays of a proposal we cannot
```

Comment:
Why cant we? Cant we access the Voting storage? I know it would involve a db read, but I think since thid db read is already made in the vote extrinsic, we can do it without charging for it after the dispatch succeeds

Reply:
Ah I see, the fields in the struct are private :(